File: node/src/chaincodeeventsbuilder.ts
/*
 * Copyright 2021 IBM All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { gateway } from '@hyperledger/fabric-protos';
import { ChaincodeEventsRequest, ChaincodeEventsRequestImpl } from './chaincodeeventsrequest';
import { GatewayClient } from './client';
import { EventsBuilder, EventsOptions } from './eventsbuilder';
import { SigningIdentity } from './signingidentity';

/**
 * Options used when requesting chaincode events.
 */
export type ChaincodeEventsOptions = EventsOptions;

export interface ChaincodeEventsBuilderOptions extends EventsOptions {
    client: GatewayClient;
    signingIdentity: SigningIdentity;
    channelName: string;
    chaincodeName: string;
}

export class ChaincodeEventsBuilder {
    readonly #options: Readonly<ChaincodeEventsBuilderOptions>;
    readonly #eventsBuilder: EventsBuilder;

    constructor(options: Readonly<ChaincodeEventsBuilderOptions>) {
        this.#options = options;
        this.#eventsBuilder = new EventsBuilder(options);
    }

    build(): ChaincodeEventsRequest {
        return new ChaincodeEventsRequestImpl({
            client: this.#options.client,
            signingIdentity: this.#options.signingIdentity,
            signedRequest: this.#newSignedChaincodeEventsRequestProto(),
        });
    }

    #newChaincodeEventsRequestProto(): gateway.ChaincodeEventsRequest {
        const result = new gateway.ChaincodeEventsRequest();
        result.setChannelId(this.#options.channelName);
        result.setChaincodeId(this.#options.chaincodeName);
        result.setIdentity(this.#options.signingIdentity.getCreator());
        result.setStartPosition(this.#eventsBuilder.getStartPosition());
        const transactionId = this.#options.checkpoint?.getTransactionId();
        if (transactionId) {
            result.setAfterTransactionId(transactionId);
        }
        return result;
    }

    #newSignedChaincodeEventsRequestProto(): gateway.SignedChaincodeEventsRequest {
        const result = new gateway.SignedChaincodeEventsRequest();
        result.setRequest(this.#newChaincodeEventsRequestProto().serializeBinary());
        return result;
    }
}

File: java/src/main/java/org/hyperledger/fabric/client/ChaincodeEventsBuilder.java
/*
 * Copyright 2021 IBM All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package org.hyperledger.fabric.client;

import com.google.protobuf.ByteString;
import org.hyperledger.fabric.protos.gateway.SignedChaincodeEventsRequest;
import java.util.Objects;

final class ChaincodeEventsBuilder implements ChaincodeEventsRequest.Builder {
    private final GatewayClient client;
    private final SigningIdentity signingIdentity;
    private final String channelName;
    private final String chaincodeName;
    private final StartPositionBuilder startPositionBuilder = new StartPositionBuilder();
    private String afterTransactionId;

    ChaincodeEventsBuilder(final GatewayClient client, final SigningIdentity signingIdentity, final String channelName,
                           final String chaincodeName) {
        Objects.requireNonNull(channelName, "channel name");
        Objects.requireNonNull(chaincodeName, "chaincode name");

        this.client = client;
        this.signingIdentity = signingIdentity;
        this.channelName = channelName;
        this.chaincodeName = chaincodeName;
    }

    @Override
    public ChaincodeEventsRequest.Builder startBlock(final long blockNumber) {
        startPositionBuilder.startBlock(blockNumber);
        return this;
    }

    @Override
    public ChaincodeEventsRequest.Builder checkpoint(final Checkpoint checkpoint) {
        checkpoint.getBlockNumber().ifPresent(startPositionBuilder::startBlock);
        this.afterTransactionId = checkpoint.getTransactionId().orElse(null);
        return this;
    }

    @Override
    public ChaincodeEventsRequest build() {
        SignedChaincodeEventsRequest signedRequest = newSignedChaincodeEventsRequestProto();
        return new ChaincodeEventsRequestImpl(client, signingIdentity, signedRequest);
    }

    private SignedChaincodeEventsRequest newSignedChaincodeEventsRequestProto() {
        org.hyperledger.fabric.protos.gateway.ChaincodeEventsRequest request = newChaincodeEventsRequestProto();
        return SignedChaincodeEventsRequest.newBuilder()
                .setRequest(request.toByteString())
                .build();
    }

    private org.hyperledger.fabric.protos.gateway.ChaincodeEventsRequest newChaincodeEventsRequestProto() {
        ByteString creator = ByteString.copyFrom(signingIdentity.getCreator());
        org.hyperledger.fabric.protos.gateway.ChaincodeEventsRequest.Builder builder = org.hyperledger.fabric.protos.gateway.ChaincodeEventsRequest.newBuilder()
                .setChannelId(channelName)
                .setChaincodeId(chaincodeName)
                .setIdentity(creator)
                .setStartPosition(startPositionBuilder.build());
        if (afterTransactionId != null) {
            builder.setAfterTransactionId(afterTransactionId);
        }
        return builder.build();
    }
}
